File: BOJ/17298.cpp
// 17298번_오큰수
#include <iostream>
#include <stack>
#include <memory.h>
#define MP make_pair
using namespace std;
int n, ans[1000001];
stack<pair<int, int> > st;
int main(){
    cin >> n;
    memset(ans, -1, sizeof(ans));
    for(int i=0; i<n; ++i){
        int val;
        cin >> val;
        if(st.empty()){
            st.push(MP(val, i));
        }
        else{
            while(val > st.top().first){ // 본인보다 큰 수 나올때까지 스택 확인
                ans[st.top().second] = val; // 현재 스택 top에 있는 원소의 NGE 저장
                st.pop();
                if(st.empty()) break; // 전부 빠지면 탈출
            }
            st.push(MP(val, i)); // 본인보다 작은 원소의 NGE 모두 저장한 뒤 push
        }
    }
    for(int i=0; i<n; ++i) cout << ans[i] << " ";
    return 0;
}

File: BOJ/17298.py
# 17298번_오큰수
import sys
input = sys.stdin.readline
n = int(input())
a = list(map(int, input().split()))

# 스택을 이용해서 NGE를 구함
# 스택 top보다 큰 값이 나오면 오큰수로 판단하고, 스택에서 삭제
NGE = [-1] * n
st = []
for i in range(n):
    # 스택 top의 NGE를 찾으면 pop
    while st and (a[st[-1]]<a[i]):
        NGE[st.pop()] = a[i]
    st.append(i) # NGE idx를 스택에 삽입

for i in NGE:
    print(i, end = ' ')
